source/common/common/logger.cc

```diff
@@ -23,45 +24,67 @@ Logger::Logger(const std::string& name) {
   logger_->flush_on(spdlog::level::critical);
 }
 
-void LockingStderrOrFileSink::logToStdErr() { log_file_.reset(); }
+SinkDelegate::SinkDelegate(DelegatingLogSinkPtr log_sink)
+    : previous_delegate_(log_sink->delegate()), log_sink_(log_sink) {
+  log_sink->setDelegate(this);
+}
 
-void LockingStderrOrFileSink::logToFile(const std::string& log_path,
-                                        AccessLog::AccessLogManager& log_manager) {
-  log_file_ = log_manager.createAccessLog(log_path);
+SinkDelegate::~SinkDelegate() {
+  assert(log_sink_->delegate() == this); // Ensures stacked allocation of delegates.
+  log_sink_->setDelegate(previous_delegate_);
 }
 
-std::vector<Logger>& Registry::allLoggers() {
-  static std::vector<Logger>* all_loggers =
-      new std::vector<Logger>({ALL_LOGGER_IDS(GENERATE_LOGGER)});
-  return *all_loggers;
+FileSinkDelegate::FileSinkDelegate(const std::string& log_path,
+                                   AccessLog::AccessLogManager& log_manager,
+                                   DelegatingLogSinkPtr log_sink)
+    : SinkDelegate(log_sink), log_file_(log_manager.createAccessLog(log_path)) {}
+
+void FileSinkDelegate::log(absl::string_view msg) {
+  // Logfiles have internal locking to ensure serial, non-interleaved
+  // writes, so no additional locking needed here.
+  //
+  // TODO(jmarantz): filesystem file write should take string_view. Adding that requires
+  // finding a way to make that work for mocked filesystem writes.
+  log_file_->write(std::string(msg));
```

Comment:
This is resolved in https://github.com/envoyproxy/envoy/pull/2916 -- the second PR to go in should eliminate this conversion.

Reply:
cleaned up now.